python/dex/interop/jax.py

```diff
@@ -189,6 +192,206 @@ def dex_call_batched(batched_args, batched_dims, func_atom):
+  return (
+      primitive(func_atom)(*arg_values),
+      primitive(evaluate_linearized)(*arg_values, *tangents_no_zeros),
+  )
+
+jax.interpreters.ad.primitive_jvps[dex_call_p] = dex_call_jvp
+
+
+# === transpose ===
+
+# TODO: Refactor this to use Dex -> Python tuple result types when these are
+# available.
+
+def dex_call_evaluate_linearized_transpose(cotangents, *args, func_atom):
+  """Evaluates the transpose of a function atom.
+
+  The function atom in question is assumed to be of the form of
+  `evaluate_linearized` from `dex_call_jvp` applied to a primal function
+  atom, with the primal arguments as constants, and the tangents being some
+  combination of constants and linear inputs.
+  """
+  assert len(args) % 2 == 0
+  num_primals = len(args) // 2
+
+  primals, tangents = args[:num_primals], args[num_primals:]
+
+  # Helper functions to build strings of primal and tangent inputs.
+  def arg_string(prefix, index_set):
+    return " ".join(f"{prefix}{i}" for i in index_set)
+
+  def tuple_string(prefix, index_set):
+    return "(" + ", ".join(f"{prefix}{i}" for i in index_set) + ")"
+
+
+  if any(jax.ad.is_undefined_primal(p) for p in primals):
+    raise RuntimeError("Primal inputs to transpose primitive are undefined.")
+
+  # Add func_atom as a variable `linearized` in the context.
+  env = api.insert(func_atom.module, api.as_cstr("linearized"), func_atom)
+
+  tangent_input_indices = [
+      i for i, t in enumerate(tangents) if jax.ad.is_undefined_primal(t)
+  ]
+  tangent_constant_indices = [
+      i for i, t in enumerate(tangents) if not jax.ad.is_undefined_primal(t)
+  ]
+
+  if len(tangent_input_indices) == 0:
+    return (None,) * len(args)
+
+  # Form a lambda which partially evaluates `linearized` at the constant primal
+  # and tangent values, with the remaining arguments (which are
+  # `UndefinedPrimal`s / linear inputs) combined into a tuple, and
+  # then transpose the lambda.
+  #
+  # For a three-input primal function with constant input for the tangent
+  # parameter at index 1, the evaluated string should look like:
+  # ```
+  # \ x0 x1 x2 u1 ct.
+  #   transposeLinear (\(t0, t2). linearized x0 x1 x2 t0 u1 t2) ct
+  # ```
+
+  # x0 x1 x2 u1 ct
+  transposed_atom_params = (
+      arg_string("x", range(num_primals)) + " " +
+      arg_string("u", tangent_constant_indices) + " ct")
+
+  # (t0, t2)
+  linear_lambda_params = tuple_string("t", tangent_input_indices)
+
+  # t0 u1 t2
+  linearized_tangent_inputs = (" ".join(
+      f"t{i}" if jax.ad.is_undefined_primal(t) else f"u{i}"
+      for i, t in enumerate(tangents)))
+
+  # x0 x1 x2 t0 u1 t2
+  linearized_inputs = (
+      arg_string("x", range(num_primals)) + " " + linearized_tangent_inputs)
+
+  # \ x0 x1 x2 u1 ct.
+  #   transposeLinear (\(t0, t2). linearized x0 x1 x2 t0 u1 t2) ct
+  transposed = eval(
+      r"\ " + transposed_atom_params + ". transposeLinear (\ " +
+      linear_lambda_params + ". linearized " + linearized_inputs + ") ct",
```

Comment:
Could you use fstrings instead of string concatenation? That would make this code so much more readable.